### PR TITLE
[FSSS-108] Feat: Add Skeleton Loading

### PR DESCRIPTION
**Pull request by lucasfp13**
## What's the purpose of this pull request?

This PR intends to add Skeleton Loading components to be visible while the real components are being loaded.

The Skeleton components will be used as a fallback for lazy-loaded components or while data is not ready.

## How does it work?

We can observe the behavior when accessing a page. As long as the data is not ready, the skeleton components will be visible.

_Desktop version:_

Screen Shot 2022-02-09 at 17 25 42
Screen Shot 2022-02-09 at 17 25 55
Screen Shot 2022-02-09 at 17 26 59

_Mobile version:_

Screen Shot 2022-02-10 at 01 06 09Screen Shot 2022-02-10 at 01 05 07
Screen Shot 2022-02-10 at 01 05 21

## References

- Related to offload the animations to the GPU: [Optimizing Repaints](https://www.afasterweb.com/2017/07/27/optimizing-repaints/);

## Checklist

- [x] CHANGELOG entry added

### CODE REVIEW

File: src/components/skeletons/ElementSkeleton/ElementSkeleton.tsx
Comment (hellofanny) [inline src/components/skeletons/ElementSkeleton/ElementSkeleton.tsx:14]: How about calling it `SkeletonElement`? It looks more like the pattern we have been following 🤔

Comment (lucasfp13) [reply src/components/skeletons/ElementSkeleton/ElementSkeleton.tsx:14]: I agree with your suggestion, I'll rename all of them

File: src/components/sections/ProductTiles/ProductTiles.tsx
Comment (eduardoformiga) [inline src/components/sections/ProductTiles/ProductTiles.tsx:11]:
```suggestion
const SKELETON_PRODUCTS: Readonly<JSX.Element[]> = [...new Array(3)].map((_, index)=> (<ProductTileSkeleton key={index} tileIndex={index + 1} variant="horizontal" />));
```
The same here regarding the 3 (array size).

File: src/components/skeletons/ElementSkeleton/element-skeleton.scss
Comment (hellofanny) [reply src/components/skeletons/ElementSkeleton/element-skeleton.scss:9]: @renatamottam what do you think? Or shall we add a new token for this?

File: src/components/skeletons/FilterSkeleton/filter-skeleton.scss
Comment (hellofanny) [inline src/components/skeletons/FilterSkeleton/filter-skeleton.scss:19]:
```suggestion
      min-height: rem(60px);
```

Comment (hellofanny) [inline src/components/skeletons/FilterSkeleton/filter-skeleton.scss:19]:
```suggestion
      min-height: var(--space-9);
```

File: src/components/skeletons/ElementSkeleton/element-skeleton.scss
Comment (hellofanny) [inline src/components/skeletons/ElementSkeleton/element-skeleton.scss:30]:
```suggestion
      height: var(--space-5);
```

File: src/components/sections/ProductGallery/product-gallery.scss
Comment (hellofanny) [inline src/components/sections/ProductGallery/product-gallery.scss:32]:
```suggestion
    min-width: rem(225px);
```

File: src/components/skeletons/ProductCardSkeleton/ProductCardSkeleton.tsx
Comment (hellofanny) [inline src/components/skeletons/ProductCardSkeleton/ProductCardSkeleton.tsx:9]: Hm, trying to think a better name for this prop. It means card inside sections like “Just Arrived”, “Deals and Promotions”, right?

maybe `groupedInSection`? 🙈

File: src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss
Comment (hellofanny) [inline src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss:1]:
```suggestion
@import "../../../styles/scaffold";
```

File: src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss
Comment (hellofanny) [inline src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss:35]:
```suggestion
        height: rem(195px);
```

File: src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss
Comment (hellofanny) [inline src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss:45]: Do we need to overwrite `height: 195px` here? 👀

File: src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss
Comment (hellofanny) [inline src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss:62]:
```suggestion
        min-height: rem(30px);
```

File: src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss
Comment (eduardoformiga) [inline src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss:16]:
```suggestion
  &[data-bordered="true"] { border: var(--border-width-0) solid var(--color-border-display); }
```

File: src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss
Comment (eduardoformiga) [inline src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss:1]:
```suggestion
@import "../../../styles/scaffold";
```
so that we can use the utilities as well

File: src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss
Comment (eduardoformiga) [inline src/components/skeletons/ProductCardSkeleton/product-card-skeleton.scss:62]:
```suggestion
        min-height: rem(30px);
```
or var(--space-5) = 32px

File: src/components/skeletons/Shimmer/shimmer.scss
Comment (eduardoformiga) [inline src/components/skeletons/Shimmer/shimmer.scss:16]:
```suggestion
    box-shadow: 0 0 var(--space-5) var(--space-5) rgb(255 255 255 / 20%);
    transform: skewX(-20deg);
```
Stylelint complains about this order

File: src/components/skeletons/Shimmer/shimmer.scss
Comment (eduardoformiga) [inline src/components/skeletons/Shimmer/shimmer.scss:14]:
```suggestion
    background: var(--color-white-transparent-20);
```
@hellofanny just notice that the `--color-white-transparent-20` is equals to `--color-black-transparent-20`. Should we change there to use white value? like `rgb(255 255 255 / 20%)`

**Comment by eduardoformiga** [comment]
Just notice now. Could you please remove it from the `product-gallery.scss`?
```
// Check it after apply Skeletons.
.product-listing__data-loading {
```

**Comment by hellofanny** [comment]
> Just notice now. Could you please remove it from the `product-gallery.scss` and update the `CHANGELOG.md`?
> 
> ```
> // Check it after apply Skeletons.
> .product-listing__data-loading {
> ```

Oh yeah! Nice cathingggg 🚀

File: src/components/skeletons/Shimmer/shimmer.scss
Comment (hellofanny) [inline src/components/skeletons/Shimmer/shimmer.scss:2]: Do you mean that the parent that use the shimmer, should have `position: relative` and `overflow: hidden` in order make it works as expected?

**Review by danzanzini — APPROVED** [review]
Great job. It's working very well. 🚀

I left some comments and questions on the code, but it's already approved.

Food for thought: I think we can extract the `loading` logic to a functional component so we can send it to Faststore but we may deal with this later.

Again, great job 🪨

File: src/components/sections/ProductGallery/ProductGallery.tsx
Comment (eduardoformiga) [reply src/components/sections/ProductGallery/ProductGallery.tsx:44]: Agree with @danzanzini. The first approach looks better to me.

File: src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx:10]:
```suggestion
```

File: src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx:13]:
```suggestion
function SkeletonProductGrid({ children, loading = true }: PropsWithChildren<Props>) {
```

File: src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx:23]:
```suggestion
    children
```

File: src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx:16]: That's kinda tricky, why not create and use an iterator?

Comment (igorbrasileiro) [reply src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx:16]: Also, would be better if it receives a property called quantityItens

File: src/components/skeletons/SkeletonProductShelf/SkeletonProductShelf.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductShelf/SkeletonProductShelf.tsx:10]:
```suggestion
```

File: src/components/skeletons/SkeletonProductShelf/SkeletonProductShelf.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductShelf/SkeletonProductShelf.tsx:13]:
```suggestion
function SkeletonProductShelf({ children, loading = true }: PropsWithChildren<Props>) {
```

File: src/components/skeletons/SkeletonProductShelf/SkeletonProductShelf.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductShelf/SkeletonProductShelf.tsx:23]:
```suggestion
    children
```

File: src/components/skeletons/SkeletonProductTiles/SkeletonProductTile/SkeletonProductTile.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductTiles/SkeletonProductTile/SkeletonProductTile.tsx:4]: Avoid using relative path
`src/components/skeletons/SkeletonElement`

File: src/components/skeletons/SkeletonProductTiles/SkeletonProductTiles.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductTiles/SkeletonProductTiles.tsx:8]:
```suggestion
```

File: src/components/skeletons/SkeletonProductTiles/SkeletonProductTiles.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductTiles/SkeletonProductTiles.tsx:26]:
```suggestion
    children
```

File: src/components/skeletons/SkeletonProductTiles/SkeletonProductTiles.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonProductTiles/SkeletonProductTiles.tsx:19]:
```suggestion
      {Array.from({ length: CONSTANT }, (_, index) => (
```
Create a constant.
Also, prefer receive itensQuantity as prop and pass this constant.

File: src/components/skeletons/SkeletonFilter/SkeletonFilter.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonFilter/SkeletonFilter.tsx:9]:
```suggestion
```

File: src/components/skeletons/SkeletonFilter/SkeletonFilter.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonFilter/SkeletonFilter.tsx:12]:
```suggestion
function SkeletonFilter({ children, loading = true }: PropsWithChildren<Props>) {
```

File: src/components/skeletons/SkeletonFilter/SkeletonFilter.tsx
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonFilter/SkeletonFilter.tsx:25]:
```suggestion
    children
```

File: src/components/skeletons/SkeletonElement/skeleton-element.scss
Comment (igorbrasileiro) [inline src/components/skeletons/SkeletonElement/skeleton-element.scss:4]: Does this prop is really necessary, depending what are you rendering, the browser will need to work twice. 
1. render and get rect for this element
2. calculate the overflow rect to paint.

File: src/components/sections/ProductGallery/ProductGallery.tsx
Comment (eduardoformiga) [inline src/components/sections/ProductGallery/ProductGallery.tsx:115]: #328 removes the loading `div`. Maybe we can wait until the merge to think of a better way to deal with it.

File: src/components/sections/ProductShelf/ProductShelf.tsx
Comment (eduardoformiga) [inline src/components/sections/ProductShelf/ProductShelf.tsx:14]:
```suggestion
  const haveProducts = products?.length > 0
```

maybe we can put it inline on line 17

File: src/components/sections/ProductTiles/ProductTiles.tsx
Comment (eduardoformiga) [inline src/components/sections/ProductTiles/ProductTiles.tsx:15]: the same here

**Review by igorbrasileiro — CHANGES_REQUESTED** [review]
Good job! Left some comments.
Does exist some reason not to use the [@faststore/ui](https://faststore.dev/reference/ui/atoms/Skeleton)?

File: src/components/skeletons/SkeletonElement/SkeletonElement.tsx
Comment (eduardoformiga) [reply src/components/skeletons/SkeletonElement/SkeletonElement.tsx:28]: is it necessary?

**Comment by igorbrasileiro** [comment]
Take a look at this reference: https://github.com/vtex/faststore/pull/558

File: src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx
Comment (eduardoformiga) [inline src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx:23]: do we need it?

File: src/components/skeletons/SkeletonProductShelf/SkeletonProductShelf.tsx
Comment (eduardoformiga) [inline src/components/skeletons/SkeletonProductShelf/SkeletonProductShelf.tsx:23]: the same here

**Review by eduardoformiga — COMMENTED** [review]
LGTM! I like the new composition approach 👏 
I left some comments

File: src/components/product/ProductGrid/ProductGrid.tsx
Comment (lucasfp13) [reply src/components/product/ProductGrid/ProductGrid.tsx:3]: I agree with you, we had a discussion about it [here](https://github.com/vtex-sites/base.store/pull/317#discussion_r802121739). For some components, like `ProductCard`, `ProductGrid` the pattern `<NameHere>Skeleton` sounds better but we kept a pattern for all of them.

If we could vary the position of the name "Skeleton" between the beginning or the end of the word, depending on which component it is, that would be great. What do you think?

File: src/components/sections/ProductGallery/ProductGallery.tsx
Comment (lucasfp13) [reply src/components/sections/ProductGallery/ProductGallery.tsx:44]: for this case, I think that is better to keep the skeleton outside because the Filter itself already has too much logic involving the `orderedFacets` results

File: src/components/skeletons/SkeletonElement/SkeletonElement.tsx
Comment (lucasfp13) [reply src/components/skeletons/SkeletonElement/SkeletonElement.tsx:28]: that was for the cases when `children` don't exist, but it'll be not necessary anymore. I'll use `PropsWithChildren`

File: src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx
Comment (lucasfp13) [reply src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx:13]: when I tried to use the type `PropsWithChildren` it raised an error when using the component as wrapper:

Screen Shot 2022-02-16 at 23 13 15

that's why I was defining `children` as a JSX.Element. Do you have any thoughts about how we handle that?

File: src/components/skeletons/SkeletonElement/skeleton-element.scss
Comment (lucasfp13) [reply src/components/skeletons/SkeletonElement/skeleton-element.scss:4]: This prop is necessary to make the `SkeletonElement` work independently and to have the shimmer effect, without needing a parent wrapper component with relative position

File: src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx
Comment (lucasfp13) [reply src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx:13]: I solve that by wrapping `children` into a fragment since a component needs to return a single element.

**Review by eduardoformiga — APPROVED** [review]
Uhuuuu 🎉 🎉 

Awesome Job!

File: src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx
Comment (igorbrasileiro) [reply src/components/skeletons/SkeletonProductGrid/SkeletonProductGrid.tsx]: About the Array.from it's ok for me.  About the quantity items, I try to explain better.

```tsx
const ProductShelf = ({ ..., pageSize }) => {
   <ProductGridSkeleton slots={pageSize}> // I think slots is better than quantityItems, idk, feel free to choose a better name
     {/* Children */}
   </ProductSkeletonGrid>
}
```
The ProductShelf component has the responsibility to know how many items he will render, right? This value comes from the SearchProvider (ITEMS_PER_PAGE)
As far as I understand, the SkeletonProductGrid, should render `a grid with slots`, how many? Idk, doesn't matter.
So, knowing this,
```tsx
const ProductGridSkeleton = ({ slots }) => {
  return <ul>{Array.from({ length: slots }, (_, index) => (<ProductCardSkeleton />))}
}
```

Does it make sense? Can go further doing this:

```tsx
const ProductShelf = ({pageSize, products, ....}) => {
  const loading = products?.length === 0
  
  return loading ? (<ProductGridSkeleton key="product-grid-skeleton" slots={pageSize} />) : (<ul key="product-grid"> {products.map(...)}</ul>)
}
```

File: src/components/skeletons/SkeletonElement/skeleton-element.scss
Comment (igorbrasileiro) [reply src/components/skeletons/SkeletonElement/skeleton-element.scss:4]: Ok, np

**Review by igorbrasileiro — APPROVED** [review]

File: CHANGELOG.md
Comment (hellofanny) [inline CHANGELOG.md:12]:
```suggestion
- `Skeleton` loading components.
```

File: src/components/skeletons/FilterSkeleton/filter-skeleton.scss
Comment (hellofanny) [inline src/components/skeletons/FilterSkeleton/filter-skeleton.scss:1]:
```suggestion
@import "src/styles/scaffold";
```

File: src/components/skeletons/ElementSkeleton/element-skeleton.scss
Comment (renatamottam) [reply src/components/skeletons/ElementSkeleton/element-skeleton.scss]: @lucasfp13 i think we should create a new token for the skeleton: `--bg-skeleton: var(--bg-disabled);`

**Review by filipewl — APPROVED** [review]
Reviewing the [files without showing whitespace](https://github.com/vtex-sites/base.store/pull/317/files?w=1) helped a lot to remove the diff clutter since it was mostly re-indentations! The skeletons were structured in a `Suspense`-way, very nice.

I feel like there was a lot of work involved with individual skeletons and new skeleton CSS files that could be a burden to maintain in tandem with their real components counterparts. But maybe we won't have to worry about it since most of this work is done now.

Great job!!! Now I want to slow down the network speed just so I can appreciate these skeletons, haha.
😆

---

Checking [LH's PLP comparison](http://lighthouse-ci.staging-2b.ingress.vtex.io/app/projects/base.store/compare/4a0e152e-f100-4a61-b2ed-d2eb358ed266?baseUrl=%2Foffice&compareUrl=%2Foffice), it seems this PR reduces the CLS from 0.0004 to 0 (no more "Avoid large layout shift" for 5 elements in the Diagnostics tab).
👏

File: src/components/skeletons/Shimmer/Shimmer.tsx
Comment (tlgimenes) [inline src/components/skeletons/Shimmer/Shimmer.tsx:5]: please use functions for components so React devTools print the component's name

File: src/components/skeletons/SkeletonElement/SkeletonElement.tsx
Comment (tlgimenes) [inline src/components/skeletons/SkeletonElement/SkeletonElement.tsx:19]: do we need to pass the loading state in here? Isn't it better to do this if outside of the component?

Comment (lucasfp13) [reply src/components/skeletons/SkeletonElement/SkeletonElement.tsx:19]: I left that way for the cases without any `children` components, so then it'll keep rendering the skeleton

File: src/components/product/ProductGrid/ProductGrid.tsx
Comment (tlgimenes) [inline src/components/product/ProductGrid/ProductGrid.tsx:16]: I think we don't need this if in here
```suggestion
    <ProductGridSkeleton loading={products.length === 0}>
```

**Comment by hellofanny** [comment]
> > Just notice now. Could you please remove it from the `product-gallery.scss` and update the `CHANGELOG.md`?
> > ```
> > // Check it after apply Skeletons.
> > .product-listing__data-loading {
> > ```
> 
> Oh yeah! Nice cathingggg 🚀

Checked [here](https://github.com/vtex-sites/base.store/pull/317/commits/15cd92eb9e7a32ba271ad759232b0568cf34d7ce)! ✅

**Review by hellofanny — APPROVED** [review]
Well done!! 🚀 🚀 What a huge PR 🙈

**Pull request merged**
